src/Interpreters/inplaceBlockConversions.cpp

```diff
@@ -43,22 +52,31 @@ void addDefaultRequiredExpressionsRecursively(const Block & block, const String
         RequiredSourceColumnsVisitor(columns_context).visit(column_default_expr);
         NameSet required_columns_names = columns_context.requiredColumns();
 
-        auto cast_func = makeASTFunction("CAST", column_default_expr, std::make_shared<ASTLiteral>(columns.get(required_column).type->getName()));
-        default_expr_list_accum->children.emplace_back(setAlias(cast_func, required_column));
-        added_columns.emplace(required_column);
-
-        for (const auto & required_column_name : required_columns_names)
-            addDefaultRequiredExpressionsRecursively(block, required_column_name, columns, default_expr_list_accum, added_columns);
+        if (is_column_in_query && convert_null_to_default)
+        {
+            auto null_as_default_expr = makeASTFunction("ifNull", std::make_shared<ASTIdentifier>(required_column_name), column_default_expr);
```

Comment:
I think that we also need to add `CAST` here for `column_default_expr`. Similar as in `else` part